### ISSUE TITLE
Bad type inference from return type

### ISSUE THREAD

**Issue by dawidgarus** [Microsoft/TypeScript#16099]
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
function flatMap<ItemType, ProjectedItemType>(projectorFn: (item: ItemType) => ProjectedItemType[]): (aggregate: ProjectedItemType[], item: ItemType) => ProjectedItemType[] {
	return (aggregate: ProjectedItemType[], current: ItemType) => aggregate.concat(projectorFn(current) || []);
}

['alice', 'bob'].reduce(flatMap(name => name.split('')), []);
['alice', 'bob'].reduce(flatMap<string, string>(name => name.split('')), []); // this works fine
```

**Expected behavior:**
TypeScript compiler should infer that type of `name` is `string`.

**Actual behavior:**
Compiler error: `Property 'split' does not exist on type '{}'.`

**Comment by ikatyang** [comment]
You have to provide some information for TS to infer types, otherwise it wouldn't know what `name` is.

for example

```ts
// <ItemType extends string>

declare function flatMap<ItemType extends string, ProjectedItemType>(projectorFn: (item: ItemType) => ProjectedItemType[]): (aggregate: ProjectedItemType[], item: ItemType) => ProjectedItemType[];

['alice', 'bob'].reduce(flatMap(name => name.split('')), []);
```

or

```ts
// (name: string) => name.split('')

declare function flatMap<ItemType, ProjectedItemType>(projectorFn: (item: ItemType) => ProjectedItemType[]): (aggregate: ProjectedItemType[], item: ItemType) => ProjectedItemType[];

['alice', 'bob'].reduce(flatMap((name: string) => name.split('')), []);
```

**Comment by dawidgarus** [comment]
@ikatyang I know, but it would be nice if compiler could work out type of name from return type.
`reduce` expects `(something[], string): something[]` function and `flatMap` returns `(ProjectedItemType[], ItemType): ProjectedItemType[]`. When compiler doesn't know what is `ItemType` as assumes that it's `{}`, it could use that knowledge to infer that `ItemType` is `string`.

**Comment by ahejlsberg** [comment]
Duplicate of #15680. Fixed by #16072.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
